### PR TITLE
Display subcategory products in parent category view

### LINKED ISSUE #1776: Difficulties customizing shop for cash on delivery
(Please take the time and describe the problem you're having before you propose a solution.)

### What I'm trying to achieve

I have a shop with the following methods of payment: card, cash on delivery and bank transfer. All use the same delivery service provider, so there's a fixed delivery cost, but when using the "cash on delivery" option, there's an additional cost. To make it easy for me, I have created a new product "Cash on delivery", so that it gets included in the bill and receipts and is transparent to the users what happens. This is implemented as a new "Payment Provider", that looks up the CashOnDelivery product and adds it to the order lines, when processing.

The problems:

* This product initially had no variant (variants didn't make sense). I can't add it from the "Add variant to order" option in the Dashboard.
* I have added a variant, but now I needed to customize 2 views, because this product is not published. Publishing it will just place its category in the main storefront navigation and that's not what I want.

Some other stuff I've noticed: 

* there's no easy "upload your logo" page, I need to customize at least the base.html and invoice.html.
* empty categories will show up on the navigation, even if they don't have published products inside. Solved by other customizations.
* category_index is not tested with subcategories. At some point at the beginning of this year, the categories have been refactored and I can assign only one category to a product. Main category, fine. But now all top categories are empty, because all products have moved to the sub-categories. The proper code should be something like this in category_index():

```python
    categories = category.get_descendants(include_self=True)
    ids = [c.id for c in categories]
    products = products_with_details(user=request.user).filter(
        category__id__in=ids).order_by('name')
    product_filter = ProductCategoryFilter(request.GET, queryset=products,
                                           category=category)
```

### PR DESCRIPTION
**Pull request by akjanik**
I want to merge this change because...

Closes #1776, #1802 

### Pull Request Checklist

(Please keep this section. It will make maintainer's life easier.)

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] Python code quality checks pass: `pycodestyle`, `pydocstyle`, `pylint`.
1. [x] JavaScript code quality checks pass: `eslint`.

### CODE REVIEW

File: saleor/product/views.py [hunk 1]
```diff
@@ -113,8 +113,11 @@ def category_index(request, path, category_id):
     if actual_path != path:
         return redirect('product:category', permanent=True, path=actual_path,
                         category_id=category_id)
+    # Check for subcategories
+    categories = category.get_descendants(include_self=True)
+    ids = [c.id for c in categories]
```
Comment (Pacu2) [inline saleor/product/views.py]: Can we use flat `values_list` on the queryset?

Comment (akjanik) [reply saleor/product/views.py]: Yes, we can!

File: tests/conftest.py [hunk 2]
```diff
@@ -642,3 +642,16 @@ def group_data(orders, status):
 def collection(db):
     collection = Collection.objects.create(name='Collection', slug='collection')
     return collection
+
+
+@pytest.fixture
+def product(product_type, default_category):
```
Comment (elwoodxblues) [inline tests/conftest.py]: We already have the `product_in_stock` fixture that creates a product. Doesn't it work for your case?

Comment (akjanik) [reply tests/conftest.py]: It works, missed that one.

**Pull request merged**